The test suite errors are:
```
The following packages have unmet dependencies:
 g++-9-multilib : Depends: gcc-9-multilib (= 9.3.0-21ubuntu1~16.04) but it is not going to be installed
                  Depends: lib32stdc++-9-dev (= 9.3.0-21ubuntu1~16.04) but it is not going to be installed
                  Depends: libx32stdc++-9-dev (= 9.3.0-21ubuntu1~16.04) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
Error: Process completed with exit code 100.
```
which appear to have nothing to do with this PR.

--- a/src/dmd/e2ir.d
+++ b/src/dmd/e2ir.d
@@ -3627,13 +3627,38 @@ elem *toElem(Expression e, IRState *irs)
                 eright = el_combine(incUsageElem(irs, ce.e2.loc), eright);
 
             tym_t ty = eleft.Ety;
+            if (tybasic(ty) == TYnoreturn)
+                ty = eright.Ety;
             if (ce.e1.type.toBasetype().ty == Tvoid ||
                 ce.e2.type.toBasetype().ty == Tvoid)
                 ty = TYvoid;
 
-            elem *e = el_bin(OPcond, ty, ec, el_bin(OPcolon, ty, eleft, eright));
-            if (tybasic(ty) == TYstruct)
-                e.ET = Type_toCtype(ce.e1.type);
+            elem* e;
+            if (tybasic(eleft.Ety) == TYnoreturn &&
+                tybasic(eright.Ety) != TYnoreturn)
+            {
+                /* ec ? eleft : eright => (ec && eleft),eright
+                 */
+                e = el_bin(OPandand, TYvoid, ec, eleft);
+                e = el_combine(e, eright);
+                if (tybasic(ty) == TYstruct)
+                    e.ET = Type_toCtype(ce.e2.type);
+            }
+            else if (tybasic(eright.Ety) == TYnoreturn)
+            {
+                /* ec ? eleft : eright => (ec || eright),eleft
+                 */
+                e = el_bin(OPoror, TYvoid, ec, eright);
+                e = el_combine(e, eleft);
+                if (tybasic(ty) == TYstruct)
+                    e.ET = Type_toCtype(ce.e1.type);
+            }
+            else
+            {
+                e = el_bin(OPcond, ty, ec, el_bin(OPcolon, ty, eleft, eright));
+                if (tybasic(ty) == TYstruct)
+                    e.ET = Type_toCtype(ce.e1.type);
+            }
             elem_setLoc(e, ce.loc);
             result = e;
         }

--- a/src/dmd/expressionsem.d
+++ b/src/dmd/expressionsem.d
@@ -11500,11 +11500,19 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
 
         Type t1 = exp.e1.type;
         Type t2 = exp.e2.type;
+        if (t1.ty == Tnoreturn)
+        {
+            exp.type = t2;
+        }
+        else if (t2.ty == Tnoreturn)
+        {
+            exp.type = t1;
+        }
         // If either operand is void the result is void, we have to cast both
         // the expression to void so that we explicitly discard the expression
         // value if any
         // https://issues.dlang.org/show_bug.cgi?id=16598
-        if (t1.ty == Tvoid || t2.ty == Tvoid)
+        else if (t1.ty == Tvoid || t2.ty == Tvoid)
         {
             exp.type = Type.tvoid;
             exp.e1 = exp.e1.castTo(sc, exp.type);

--- a/test/runnable/noreturn1.d
+++ b/test/runnable/noreturn1.d
@@ -1,9 +1,8 @@
-
-/*****************************************/
-
 alias noreturn = typeof(*null);
 
 extern (C) noreturn exit();
+
+/*****************************************/
 
 bool testf(int i)
 {
@@ -15,28 +14,61 @@ bool testt(int i)
     return i || assert(0);
 }
 
-int test3(int i)
+int testa(int i)
 {
     if (i && exit())
         return i + 1;
     return i - 1;
 }
 
-int test4(int i)
+int testb(int i)
 {
     if (i || exit())
         return i + 1;
     return i - 1;
 }
 
-int main()
+void test1()
 {
     assert(testf(0) == false);
     assert(testt(1) == true);
 
-    assert(test3(0) == -1);
-    assert(test4(3) == 4);
-
-    return 0;
+    assert(testa(0) == -1);
+    assert(testb(3) == 4);
 }
 
+/*****************************************/
+
+noreturn exit1() { assert(0); }
+noreturn exit2() { assert(0); }
+
+
+int heli1(int i)
+{
+    return i ? exit1() : i - 1;
+}
+
+int heli2(int i)
+{
+    return i ? i - 1 : exit1();
+}
+
+noreturn heli3(int i)
+{
+    return i ? exit1() : exit2();
+}
+
+void test2()
+{
+    assert(heli1(0) == -1);
+    assert(heli2(1) == 0);
+}
+
+/*****************************************/
+
+int main()
+{
+    test1();
+    test2();
+    return 0;
+}

